src/components/tooltip/CallNode.js

```diff
@@ -57,15 +62,241 @@ type Props = {|
+    maybeDisplayData: ?CallNodeDisplayData
+  ) {
+    if (!maybeTimings || !maybeDisplayData) {
+      return null;
+    }
+    const { totalTime, selfTime } = maybeTimings.forPath;
+    if (!totalTime.breakdownByCategory) {
+      return null;
+    }
+
+    const totalTimeBreakdownByCategory: BreakdownByCategory =
+      totalTime.breakdownByCategory;
+
+    const { thread, weightType, categories } = this.props;
+
+    // JS Tracer threads have data relevant to the microsecond level.
+    const isHighPrecision = Boolean(thread.isJsTracer);
+
+    const categoriesAndTime: Array<{
+      categoryName: string,
+      category: number,
+      subCategory: number, // -1 for category headers
+      totalTime: number,
+      selfTime: number,
+    }> = [];
+    totalTimeBreakdownByCategory.forEach(
+      ({ entireCategoryValue, subcategoryBreakdown }, category) => {
+        if (entireCategoryValue === 0) {
+          return;
+        }
+        const selfTimeValue = selfTime.breakdownByCategory
+          ? selfTime.breakdownByCategory[category].entireCategoryValue
+          : 0;
+
+        const filteredSubcategoryBreakdown: Array<{
+          subcategory: number,
+          self: number,
+        }> = subcategoryBreakdown
+          .map((subval, i) => {
+            return { subcategory: i, self: subval };
+          })
+          .filter((sub) => sub.self !== 0);
+        if (filteredSubcategoryBreakdown.length === 1) {
+          const { subcategory } = filteredSubcategoryBreakdown[0];
+          const categoryName =
+            categories[category].name +
+            (subcategory !== 0
+              ? ': ' + categories[category].subcategories[subcategory]
+              : '');
+          categoriesAndTime.push({
+            categoryName,
+            category,
+            subCategory: -1,
+            selfTime: selfTimeValue,
+            totalTime: entireCategoryValue,
+          });
+          return;
+        }
+        const categoryName = categories[category].name;
+        categoriesAndTime.push({
+          categoryName,
+          category,
+          subCategory: -1,
+          selfTime: selfTimeValue,
+          totalTime: entireCategoryValue,
+        });
+        if (filteredSubcategoryBreakdown[0].subcategory === 0) {
+          const other = filteredSubcategoryBreakdown[0];
+          filteredSubcategoryBreakdown.pop();
+          filteredSubcategoryBreakdown.push(other);
+        }
+        filteredSubcategoryBreakdown.forEach((sub) => {
+          const { self, subcategory } = sub;
+          const selfTimeValue = selfTime.breakdownByCategory
+            ? selfTime.breakdownByCategory[category].subcategoryBreakdown[
+                subcategory
+              ]
+            : 0;
+          categoriesAndTime.push({
+            categoryName,
+            category,
+            subCategory: subcategory,
+            selfTime: selfTimeValue,
+            totalTime: self,
+          });
+        });
+      }
+    );
```

Comment:
I think that can be rewritten in a simpler way. Especially the filter and all this is, I think, unnecessary.

Here is some pseudocode:

```js
function hasAtLeastTwoPositiveValues(arr: number[]): boolean {
  let count = 0;
  for (const value of arr) {
    if (value > 0) { 
      if (++count === 2) {
        return true;
      }
    }
  }
  return false;
}

totalTimeBreakdownByCategory.forEach(
  ({ entireCategoryValue, subcategoryBreakdown }, category) => {
    if (entireCategoryValue === 0) {
      return;
    }
    if (!hasAtLeastTwoPositiveValues(subcategoryBreakdown) {
      categoriesAndTime.push({
        categoryName: getCategoryPairLabel(...),
        category,
        subCategory: -1,
        selfTime: selfTimeValue,
        totalTime: entireCategoryValue,
      });
      return;
    }
    
    // It has at least 2 values
    categoriesAndTime.push({
      ...
    });
    
    // Start at 1, we'll add the "Other" subcategory at the end if needed.
    for (let subCategoryIndex = 1; subCategoryIndex < subcategoryBreakdown.length; subCategoryIndex++) {
      const subCategoryValue = subcategoryBreakdown[subCategoryIndex];
      if (subCategoryValue === 0) { continue }
      categoriesAndTime.push({
        ...
      });
    }
    
    // If the "other" subcategory is present, add it now.
    if (subcategoryBreakdown[0] > 0) {
      categoriesAndTime.push({
        ...
      });
    }
  }
);
```

Now I even think that this could work without this `categoriesAndTime` intermediate structure, by directly rendering some JSX. I can think of 3 functions:
* one function `_renderOneSubCategoryLine` that takes `categoryName, categoryIndex, subCategory, selfTime, totalTime` and renders the JSX for one line
* one function `_renderOneCategory` that renders one category and its subcategories (if applicable). This would return null if there's nothing to render, and otherwise call `_renderOneSubCategoryLine`.
* a last function that renders the full list by running the algorithm above and calling `_renderOneCategory` for each category.

Benefits: more direct code, more imperative, less intermediate structures, less if conditions, function names help document the code, easier to follow.

With these 3 functions we can also easily do one last change: move the `Other` category at the bottom, by starting the loop at index 1 (with a for loop instead of `forEach`), and add it if needed.

How does that sound? I realize this is rewriting the algorithm, but I think this will lead to more maintainable code in the longer run.

Reply:
if you think it's too much work, I think we can try to move forward with your solution, with some more comments.

Reply:
> Benefits: more direct code, more imperative, less intermediate structures, less if conditions, function names help document the code, easier to follow.

I don't know. I rewrote the data structures and moved the whole (with your nice suggestions) into a separate function which makes the rendering code fairly easy to understand.

> if you think it's too much work, I think we can try to move forward with your solution, with some more comments.

Only mixing in the JSX code would require some work and code duplication.

Reply:
I'm currently rewriting my own solution to your proposed JSX solution. You were right to propose it.